Define way to create object with async constructor

ela-kotulska-frequenz opened the pull request.

ela-kotulska-frequenz committed:
Define a way for constructing objects with async constructor

Add parent class that raises error if __init__ was called.
Use this for implementing BatteryPoolStatus object.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Make StatusTracker AsyncConstructible

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

shsms commented on src/frequenz/sdk/actor/power_distributing/_battery_pool_status.py:
```diff
@@ -4,28 +4,35 @@
-        Note:
-            Please call `async_init` method to fully initialize BatteryPoolStatus. Otherwise
-            it is not possible to use BatteryPoolStatus.
+    ) -> "BatteryPoolStatus":
```
can you use `from __future__ import annotations` instead of the quotes?  I guess there's no need to write python 3.8 friendly code anymore.

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/_battery_pool_status.py:
```diff
@@ -35,20 +42,21 @@ def __init__(
+        Returns:
+            New instance of this class.
         """
+        self: "BatteryPoolStatus" = BatteryPoolStatus.__new__(cls)
```
here too.

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -164,15 +164,9 @@ def __init__(
-        )
+        # This will be temporary initialized in the `run` method.
+        # PowerDistributingActor won't use BatteryPoolStatus soon.
+        self._battery_pool: BatteryPoolStatus
```
I see just a single use of this variable outside the `run` method.  I think it might be cleaner to make it a local variable inside the `run` function,  and pass it as argument from there.

ela-kotulska-frequenz replied:
Done ( in new commit)

shsms commented on src/frequenz/sdk/util/asyncio.py:
```diff
@@ -19,3 +20,22 @@ async def cancel_and_await(task: asyncio.Task) -> None:
+        raise NotSyncConstructible(
+            "This object shouldn't be created with default constructor. ",
+            "Check class documentation for more information.",
+        )
```
maybe also add an abstract method, to standardize all uses?

``` python
@abstractmethod
async def async_new() -> Self:
  """Create an `AsyncConstructible` object."""
```

ela-kotulska-frequenz replied:
I could do that, but the problem is that then number of constructor paramenters in async_new  method might different it he class that derives from it.

So in the `AsyncConstructible::async_new(cls: Type[T]) -> T `
But the classes that derives after it have different number of constructor arguments :confused:
And we get the pylint error:
```
Number of parameters was 1 in 'AsyncConstructible.async_new' and is now 4 in overridden 'StatusTracker.async_new' method
```
And the only way (I found) to avoid it is to create AsyncConstructibleConfig abstract class, and each class that derives from AsyncConstructible has to write its own config class that is passed to the constructor.

ela-kotulska-frequenz replied:
As discussed it stays like this for now

ela-kotulska-frequenz committed:
Make battery_pool variable local in PowerDistributingActor

Move parse_result method to set_distributed_power to decrease
nubmer of local variables.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add toml and tomli dependencies to CONTRIBUTING

toml is needed to create min_requiremenst file
tomli is needed to parse pyproject configuration file.
Without tomli I see warning:
`configuration file was ignored, because the `tomli`
package is not installed`

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>
